Add OpenCL support for sloth encode

simonatsn opened the pull request.

simonatsn committed:
Add OpenCL support for sloth encode

nazar-pc commented on benches/cuda.rs:
```diff
@@ -10,17 +11,27 @@ fn random_bytes<const BYTES: usize>() -> Vec<u8> {
-    let expanded_ivs = random_bytes::<32768>(); // 1024 * 32
+    const NUM_PIECES: usize = 1024 * 1024;
+    let size = NUM_PIECES * 4096 / 1073741824;
+    let big_piece = random_bytes::<{4096 * NUM_PIECES}>();
```
What was the reason for changing this to allocate 4G of memory? Seems like a lot to me.

simonatsn replied:
The idea is to allocate sufficient space that the GPU can be kept busy between the compute and data transfer. The 4G space is subdivided into sub-buffers (see https://github.com/simonatsn/sloth256-189/blob/master/src/opencl/opencl.cpp#L32), all of which operate in parallel. This allows compute and transfer to overlap. The use of 4G is somewhat arbitrary, ie a smaller amount is also probably sufficient to get most of the benefit. However, too small and the overhead will dominate. Our assumption is vast amounts of data will need to be encoded so larger is better.

nazar-pc commented on build.rs:
```diff
@@ -59,4 +59,28 @@ fn main() {
+
+        println!("cargo:rustc-link-lib=OpenCL");
+        println!("cargo:rustc-link-search=/opt/amdgpu-pro/lib64/");
+        println!("cargo:rustc-link-search=C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.4/lib/x64");
+        println!("cargo:rustc-link-search=C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.5/lib/x64");
```
Can't these and some of the below be handled in a nicer way? I see there is `.cuda(true)` above, is it something different?

And for Intel/AMD as well unless not supported yet.

simonatsn replied:
This is for Windows support only. Adding a .cuda(true) and compiling the OpenCL files with nvcc may be possible, with a fallback to compilation without nvcc in case CUDA is not installed.

nazar-pc commented on src/opencl/mod.rs:
```diff
@@ -0,0 +1,158 @@
+    return Ok(instances);
+}
+
+/// Cleans up the resources allocated in the initialization of the encode kernel.
+/// Calling this once at the end is sufficient.
+pub fn cleanup(instances: *const u8) -> Result<(), OpenCLEncodeError> {
+    let return_code = unsafe {
+        ffi::sloth256_189_opencl_cleanup(instances)
+    };
+
+    if return_code != 0 {
+        return Err(OpenCLEncodeError::OpenCLError(return_code));
+    }
+
+    return Ok(());
+}
```
Functions like these are inherently unsafe.

All of these functions should either be private and re-exposed as one (if there is no reason to not use pinned memory for instance) or wrapped into a struct with public API that is impossible to use incorrectly.

simonatsn replied:
That seems like a reasonable approach (creating a struct with a public API). Merging them all into a single function would make it so that initialization and cleanup is done multiple times, which is a major performance loss for data of smaller sizes.

nazar-pc commented on benches/opencl.rs:
```diff
@@ -0,0 +1,65 @@
+                sloth256_189::opencl::encode(&mut piece, &expanded_ivs, layers, instances).unwrap();
+                let elapsed = start.elapsed();
+
+                sloth256_189::opencl::pinned_memory_free(instances).unwrap();
+                std::mem::forget(piece);
```
`piece` should have been a wrapper struct with `Drop` implementation that does this automatically, this C-style de-allocation is not not idiomatic and error-prone.

simonatsn replied:
Again, makes sense as a good way to integrate it.

nazar-pc commented on benches/opencl.rs:
```diff
@@ -0,0 +1,65 @@
+
+            b.iter_batched_ref(
+                || piece.clone(),
+                |mut input| sloth256_189::opencl::encode(&mut input, &expanded_ivs, layers, instances).unwrap(),
```
Is there a good reason to not use pinned memory at all?

ozgunozerk replied:
Found this in: https://forums.developer.nvidia.com/t/advantages-disadvantages-of-using-pinned-memory/34422

> The only real drawbacks to pinning memory are the reduction in available physical ram to the host demand-paging system, and the time it takes to pin memory (which is significantly longer than an ordinary malloc of the same size).

Also, I remember that pinned memory can't be pre-empted.

ozgunozerk replied:
So whether it should be the `default` for our case, I would like to hear your opinions as well @simonatsn :)

simonatsn replied:
The main reason to consider not using pinned memory is based on some limitations in OpenCL. Nvidia's support for OpenCL limits how much space you can allocate for pinned memory (~1/3 of GPU memory) and it allocates twice that much on the GPU at runtime. So half of that is wasted. Other than that, for performance reasons I don't think there is any reason you would not want to use it. The approach taken in this change is to allocate the buffers once so the additional overhead there is negligible.

nazar-pc commented on src/opencl/opencl.cpp:
```diff
@@ -0,0 +1,403 @@
+#define INTEL_FACTOR 1.3f
+
+extern "C"
+cl_int sloth256_189_opencl_batch_encode(unsigned char* inout, size_t total_size, unsigned char* iv, size_t layers, EncodeOpenCLInstances* instances) {
```
Explanation throughout this function is needed, I don't fully understand what is happening here.

nazar-pc commented on src/opencl/opencl.cpp:
```diff
@@ -0,0 +1,403 @@
+
+            instance->context = clCreateContext(NULL, 1, &instance->device, NULL, NULL, &err); OCL_ERR_CHECK(err, instances);
+
+            std::string encode_path = "src/opencl";
+            std::string header_str = "";
+            if (instance->platform == NVIDIA) {
+                std::string ptx_str = read_from_file(encode_path + "/mod256-189.cu", err); OCL_ERR_CHECK(err, instances);
+                header_str = ptx_str + "\n" + read_from_file(encode_path + "/nvidia_specific.cl", err); OCL_ERR_CHECK(err, instances);
+            }
+            else {
+                header_str = read_from_file(encode_path + "/non_nvidia.cl", err); OCL_ERR_CHECK(err, instances);
+            }
+            std::string kernel_str = read_from_file(encode_path + "/encode.cl", err); OCL_ERR_CHECK(err, instances);
```
I assume this will not work as files are moved around? Like executable using this library moved somewhere with no source code available anymore. If so, then source code of those should be bundled with executable (can be `include_bytes` macro in Rust and pointer sent into here for instance).

nazar-pc reviewed (commented):
I just assume we'll have to re-format code and write necessary Rust wrappers to make it actually safe to use, but we need comments explaining what happens and even more importantly why. Also CI should pass here (including OpenCL compilation and maybe even CPU run on 3 major platforms, see https://github.com/subspace/sloth256-189/blob/master/.github/workflows/ci.yaml).

ozgunozerk commented on src/opencl/nvidia_specific.cl:
```diff
@@ -0,0 +1,107 @@
+        rp[i] = ap[i];
+}
+
+inline void t_op_eq(struct vec256_t* x, const vec256 a) {
```
some basic comments for the function signatures explaining what are the functions are doing would be really helpful. For example: `t_op_eq` was not clear if it was performing some equality check or assignment at the first glance. I had to read and understand the code of `vec_copy` for that (this is just a small example, it is getting more chaotic in more complex functions)

ozgunozerk commented on src/opencl/nvidia_specific.cl:
```diff
@@ -0,0 +1,107 @@
+
+#undef sqr_n_mul
+
+    t_op_eq(&x, redc_mod_256_189_val(x).l);
+    //x = redc_mod_256_189_val(x);
```
I assume `redc_mod_256_189_val(x).l` is performing the last step for the lazy reduction, and `.l` part is the low bits. So why using `t_op_eq` is faster than the more native way of the commented out part?
Again, some comments would be really useful. Making assumptions on top of other assumptions is not very comfortable from the code-readers point of view :)

ozgunozerk commented on src/opencl/mod256-189.cu:
```diff
@@ -0,0 +1,970 @@
+
+//asm volatile(".extern .shared .b32 scratchpad[];\t\n");
+
+struct vec256_shared_t {
```
Again, some basic comments would be beneficial for an outsider to understand, or even for us to remember what is the purpose here. I assume this struct is for shared memory, v0 is storing the vector itself, and v_off is the offset, right? But if I'm wrong, it will be very hard for me to understand the rest of the code. I think it will be nice to have this explanatory comments for every struct defined

ozgunozerk committed:
potential bug fixed

ozgunozerk committed:
malloc outside the loop

ozgunozerk committed:
simplify the code by removing default 8GB start

simonatsn committed:
Additional comments and code formatting cleanup
Autodetect available CUDA versions
Forward errors from c++ to rust
Error strings for OpenCL errors
Include opencl code as a string in the binary
Reduce benchmark encode sizes to 1024 pieces and remove serial test flag

ozgunozerk committed:
suggestions addressed

nazar-pc commented:
@simonatsn can you push commit with `cargo fmt` applied so this can actually try to pass CI?

simonatsn committed:
Run cargo fmt.

nazar-pc commented:
Yeah, now clippy isn't happy :slightly_smiling_face:

simonatsn committed:
Fix clippy error

nazar-pc commented on src/opencl/opencl.cpp:
```diff
@@ -216,130 +329,276 @@ void sloth256_189_opencl_batch_encode_instance(unsigned char* inout_offset, size
 #define CPU_FACTOR 0.75f
 #define NVIDIA_FACTOR 100.0f
 #define AMD_FACTOR 15.0f
 #define INTEL_FACTOR 1.3f
```
It would be really-really nice to have some kind of function that does quick benchmark and returns these constants in form of a struct that can be later used as an argument for `sloth256_189_opencl_batch_encode`. This is something that will be harder for us to implement on our own.

nazar-pc reviewed (changes requested):
I believe code is broken now, I don't see any changes to the benchmark, tests and some other despite summary in the commit message, is this a partial commit?

It might not be feasible in your workflow, but every line in commit summary should be a separate commit just the way developer writes them for easier review rather than big code dumps that I need to analyze for missing pieces.

Some of the comments are still left unaddressed (and left unresolved) like dead code in comments.

Otherwise appreciate added comments and improved readability :+1:

simonatsn committed:
Picking up dropped changes from prior commits.

nazar-pc commented on src/opencl/mod.rs:
```diff
@@ -59,42 +64,83 @@ mod ffi {
+    let mod256_189_cu = include_bytes!("mod256-189.cu");
+    let non_nvidia_cl = include_bytes!("non_nvidia.cl");
+
+    let encode_cl_str = String::from_utf8_lossy(encode_cl).to_string() + &"\0".to_string();
```
I will refactor this, but FYI https://doc.rust-lang.org/stable/std/ffi/struct.CString.html exists and `format!()` is more commonly used instead of string concatenation will will do more allocations.

So I'd write this instead:
```rust
    let encode_cl_cstr = CString::new(include_str!("encode.cl")).expect("File doesn't contain 0 bytes");
```

When `CString::new` will become `const`, the whole thing will be processed at compile time and with 0 byte at the end.

nazar-pc reviewed (approved):
This is mergeable from my point of view, @ozgunozerk please take a look at lower-level code and try to run this on the machines you have

nazar-pc committed:
Fix CI errors, enable more targets

nazar-pc committed:
Merge pull request #17 from subspace/fix-ci

Fix CI errors, enable more targets

nazar-pc committed:
Merge branch 'master' into remaining_piece-potential-bug-fix

ozgunozerk commented on src/opencl/non_nvidia.cl:
```diff
@@ -9,43 +9,18 @@ typedef uint limb_t;
-         * mispredicted branch. Besides, some compilers unroll the
-         * loop and remove redundant stores to 'restict'-ed storage...
-         */
-        //ret[n / sizeof(limb_t)] = limb;
```
why are these comments deleted? were they misleading?

ozgunozerk commented on src/opencl/non_nvidia.cl:
```diff
@@ -9,43 +9,18 @@ typedef uint limb_t;
-        for (limb = in[i], j = 0; j < r; j++, limb >>= 8)
-            *out++ = (unsigned char)limb;
-    }*/
-
```
likewise, why are these comments deleted?

ozgunozerk commented:
On Mac M1, it worked for CPU (I was curious about OpenCL, did not work due to missing `omp.h` header)
For my NVidia GPU (GTX 1660) tests passed both for Windows10 and Ubuntu 20.04 (--feature cuda)
For my AMD GPU (Vega 56) tests failed both for Windows10 and Ubuntu 20.04 (--feature opencl)

Here is the detailed output for Windows AMD GPU:
```
---- opencl::test_random_piece stdout ----
thread 'opencl::test_random_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_INVALID_KERNEL_NAME")', tests\integration\opencl.rs:31:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- opencl::test_known_piece stdout ----
thread 'opencl::test_known_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_INVALID_KERNEL_NAME")', tests\integration\opencl.rs:54:42

---- opencl::test_random_piece_pinned stdout ----
thread 'opencl::test_random_piece_pinned' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_INVALID_KERNEL_NAME")', tests\integration\opencl.rs:70:42


failures:
    opencl::test_known_piece
    opencl::test_random_piece
    opencl::test_random_piece_pinned

test result: FAILED. 2 passed; 3 failed; 2 ignored; 0 measured; 0 filtered out; finished in 0.12s
```

And here is the detailed output for the Ubuntu AMD GPU:
```
running 7 tests
test opencl::test_big_random_piece ... ignored
test opencl::test_big_random_piece_pinned ... ignored
test cpu::test_known_piece ... ok
test opencl::test_known_piece ... FAILED
test opencl::test_random_piece_pinned ... FAILED
test opencl::test_random_piece ... FAILED
test cpu::test_random_piece ... ok

failures:

---- opencl::test_known_piece stdout ----
thread 'opencl::test_known_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("Unknown OpenCL error")', tests/integration/opencl.rs:54:42

---- opencl::test_random_piece_pinned stdout ----
thread 'opencl::test_random_piece_pinned' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("Unknown OpenCL error")', tests/integration/opencl.rs:70:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- opencl::test_random_piece stdout ----
thread 'opencl::test_random_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("Unknown OpenCL error")', tests/integration/opencl.rs:31:42


failures:
    opencl::test_known_piece
    opencl::test_random_piece
    opencl::test_random_piece_pinned

test result: FAILED. 2 passed; 3 failed; 2 ignored; 0 measured; 0 filtered out; finished in 0.21s
```

ozgunozerk commented:
OpenCL support for CUDA works (both on Windows and Ubuntu with my GTX 1660).

This is of course good news, however CUDA support was already there, and unfortunately I couldn't make this code work under Vega 56 :/

nazar-pc commented:
@simonatsn is it possible that this code uses some features only present on RDNA architecture and not applicable for older architectures, do you have the hardware to check?

ozgunozerk committed:
Merge pull request #16 from subspace/remaining_piece-potential-bug-fix

potential bug fixed

ozgunozerk commented:
Tested OpenCL on a MacBook with intel (integrated). Had to install OMP library manually.
Then I had to change this line (https://github.com/subspace/sloth256-189/pull/14/files?show-viewed-files=true&file-filters%5B%5D=#diff-c340f7be38dbac6e86e446008871b68ead9b769a0ecda22fe959d892485eeb1dR11) to:
```
#include <OpenCL/cl.h>
```
This change is suggested in the relevant place.

After these, the error I got is the following:
```
~/IdeaProjects/sloth256-189 on supra !1 ?3 ............ INT took 8s at 16:09:47
> cargo test --features=opencl
   Compiling sloth256-189 v0.2.2 (/Users/ozgunozerk/IdeaProjects/sloth256-189)
   Compiling criterion v0.3.5
The following warnings were emitted during compilation:

warning: src/opencl/opencl.cpp:50:18: warning: unused variable 'num_threads' [-Wunused-variable]
warning:     unsigned int num_threads = std::thread::hardware_concurrency();
warning:                  ^
warning: src/opencl/opencl.cpp:222:37: error: use of undeclared identifier 'CL_DEVICE_WAVEFRONT_WIDTH_AMD'
warning:                                     CL_DEVICE_WAVEFRONT_WIDTH_AMD,
warning:                                     ^
warning: src/opencl/opencl.cpp:262:35: error: use of undeclared identifier 'clCreateCommandQueueWithProperties'
warning:                 instance->cq[i] = clCreateCommandQueueWithProperties(instance->context,
warning:                                   ^
warning: 1 warning and 2 errors generated.

error: failed to run custom build command for `sloth256-189 v0.2.2 (/Users/ozgunozerk/IdeaProjects/sloth256-189)`

Caused by:
  process didn't exit successfully: `/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-467d1ffc5730aa2a/build-script-build` (exit status: 1)
  --- stdout
  TARGET = Some("x86_64-apple-darwin")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-apple-darwin")
  CC_x86_64-apple-darwin = None
  CC_x86_64_apple_darwin = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-apple-darwin = None
  CFLAGS_x86_64_apple_darwin = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CC_x86_64-apple-darwin = None
  CC_x86_64_apple_darwin = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-apple-darwin = None
  CFLAGS_x86_64_apple_darwin = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CC_x86_64-apple-darwin = None
  CC_x86_64_apple_darwin = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-apple-darwin = None
  CFLAGS_x86_64_apple_darwin = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CC_x86_64-apple-darwin = None
  CC_x86_64_apple_darwin = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-apple-darwin = None
  CFLAGS_x86_64_apple_darwin = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-arch" "x86_64" "-Wall" "-Wextra" "-mno-avx" "-fno-builtin-memcpy" "-Wno-unused-command-line-argument" "-Werror" "-o" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/src/cpu/sloth256_189.o" "-c" "src/cpu/sloth256_189.c"
  exit status: 0
  running: "cc" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-arch" "x86_64" "-Wall" "-Wextra" "-mno-avx" "-fno-builtin-memcpy" "-Wno-unused-command-line-argument" "-Werror" "-o" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/src/cpu/assembly.o" "-c" "src/cpu/assembly.S"
  exit status: 0
  AR_x86_64-apple-darwin = None
  AR_x86_64_apple_darwin = None
  HOST_AR = None
  AR = None
  running: "ar" "cq" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/libsloth256_189.a" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/src/cpu/sloth256_189.o" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/src/cpu/assembly.o"
  exit status: 0
  running: "ar" "s" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/libsloth256_189.a"
  exit status: 0
  cargo:rustc-link-lib=static=sloth256_189
  cargo:rustc-link-search=native=/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out
  cargo:rustc-link-lib=OpenCL
  cargo:rustc-link-search=/opt/amdgpu-pro/lib64/
  TARGET = Some("x86_64-apple-darwin")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-apple-darwin")
  CXX_x86_64-apple-darwin = None
  CXX_x86_64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-apple-darwin = None
  CXXFLAGS_x86_64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CXX_x86_64-apple-darwin = None
  CXX_x86_64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-apple-darwin = None
  CXXFLAGS_x86_64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CXX_x86_64-apple-darwin = None
  CXX_x86_64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-apple-darwin = None
  CXXFLAGS_x86_64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CXX_x86_64-apple-darwin = None
  CXX_x86_64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-apple-darwin = None
  CXXFLAGS_x86_64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CXX_x86_64-apple-darwin = None
  CXX_x86_64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-apple-darwin = None
  CXXFLAGS_x86_64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CXX_x86_64-apple-darwin = None
  CXX_x86_64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-apple-darwin = None
  CXXFLAGS_x86_64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  CXX_x86_64-apple-darwin = None
  CXX_x86_64_apple_darwin = None
  HOST_CXX = None
  CXX = None
  CXXFLAGS_x86_64-apple-darwin = None
  CXXFLAGS_x86_64_apple_darwin = None
  HOST_CXXFLAGS = None
  CXXFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("cmpxchg16b,fxsr,sse,sse2,sse3,ssse3")
  running: "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-arch" "x86_64" "-I" "/opt/amdgpu-pro/include/" "-I" "/opt/intel/inteloneapi/compiler/latest/linux/include/sycl/" "-I" "" "-Wall" "-Wextra" "-pthread" "-std=c++17" "-o" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/src/opencl/opencl.o" "-c" "src/opencl/opencl.cpp"
  cargo:warning=src/opencl/opencl.cpp:50:18: warning: unused variable 'num_threads' [-Wunused-variable]
  cargo:warning=    unsigned int num_threads = std::thread::hardware_concurrency();
  cargo:warning=                 ^
  cargo:warning=src/opencl/opencl.cpp:222:37: error: use of undeclared identifier 'CL_DEVICE_WAVEFRONT_WIDTH_AMD'
  cargo:warning=                                    CL_DEVICE_WAVEFRONT_WIDTH_AMD,
  cargo:warning=                                    ^
  cargo:warning=src/opencl/opencl.cpp:262:35: error: use of undeclared identifier 'clCreateCommandQueueWithProperties'
  cargo:warning=                instance->cq[i] = clCreateCommandQueueWithProperties(instance->context,
  cargo:warning=                                  ^
  cargo:warning=1 warning and 2 errors generated.
  exit status: 1

  --- stderr


  error occurred: Command "c++" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "-m64" "-arch" "x86_64" "-I" "/opt/amdgpu-pro/include/" "-I" "/opt/intel/inteloneapi/compiler/latest/linux/include/sycl/" "-I" "" "-Wall" "-Wextra" "-pthread" "-std=c++17" "-o" "/Users/ozgunozerk/IdeaProjects/sloth256-189/target/debug/build/sloth256-189-a710c65aaaf1d07f/out/src/opencl/opencl.o" "-c" "src/opencl/opencl.cpp" with args "c++" did not execute successfully (status code exit status: 1).


warning: build failed, waiting for other jobs to finish...
```

ozgunozerk commented on src/opencl/opencl_utils.hpp:
```diff
@@ -0,0 +1,242 @@
+#include <iostream>
+
+#ifdef __APPLE__
+#include <Open/opencl.h>
```
Shouldn't this be:
`#include <OpenCL/cl.h>`
The original version did not work on my MacBook with intel CPU. Had to make this change manually.

simonatsn committed:
Fixes for MacOS and open source drivers.

simonatsn committed:
Merge branch 'master' of https://github.com/subspace/sloth256-189

simonatsn committed:
Automatically determine performance ratio factors for various platforms using short benchmarks

nazar-pc commented on src/opencl/opencl_utils.hpp:
```diff
@@ -46,72 +48,93 @@ void getAllPlatforms(cl_platform_id*& platforms, cl_uint& num_platforms, cl_int&
 
 std::string getPlatformName(cl_platform_id& platform_id, cl_int& err) {
```
Looks like all calls of this function just throw results away, is it still needed or `getDeviceVendor` is sufficient?

nazar-pc commented on src/opencl/opencl_utils.hpp:
```diff
@@ -46,72 +48,93 @@ void getAllPlatforms(cl_platform_id*& platforms, cl_uint& num_platforms, cl_int&
     return platform_name;
 }
 
 std::string getDeviceName(cl_device_id& device_id, cl_int& err) {
```
This doesn't seem to be used either

nazar-pc reviewed (commented):
Makes sense to me

ozgunozerk commented on src/opencl/opencl.cpp:
```diff
@@ -0,0 +1,762 @@
+                case AMD:
+                    size_t amd_wavefront_size;
+                    clGetDeviceInfo(instance->device,
+                                    CL_DEVICE_WAVEFRONT_WIDTH_AMD,
```
I've run the basic opencl test on my MacBook. It works. 
However, this line creates an error when I try to build the code. It says `src/opencl/opencl.cpp:321:37: error: use of undeclared identifier 'CL_DEVICE_WAVEFRONT_WIDTH_AMD'`

ozgunozerk commented:
It still doesn't work on Linux.
Here is the output of `cargo test features=opencl -- --nocapture`:
```
unning 7 tests
test opencl::test_big_random_piece ... ignored
test opencl::test_big_random_piece_pinned ... ignored
test cpu::test_known_piece ... ok

Error while building OpenCL program:
fatal error: cannot open file '/usr/local//usr/lib/clc/gfx900-amdgcn-mesa-mesa3d.bc': No such file or directory


Error while building OpenCL program:
fatal error: cannot open file '/usr/local//usr/lib/clc/gfx900-amdgcn-mesa-mesa3d.bc': No such file or directory


Error while building OpenCL program:
thread 'fatal error: cannot open file '/usr/local//usr/lib/clc/gfx900-amdgcn-mesa-mesa3d.bc': No such file or directory
opencl::test_known_piece
' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:thread '55opencl::test_random_piece_pinned:' panicked at '42called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")
', note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
tests/integration/opencl.rs:72:thread '42
opencl::test_random_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:31:42
test opencl::test_random_piece_pinned ... FAILED
test opencl::test_known_piece ... FAILED
test opencl::test_random_piece ... FAILED
test cpu::test_random_piece ... ok
```
The path for finding the file is not correct. There is an extra `/usr/local/` there in my opinion.
If would be helpful, a similar file exists in my system in the following path: `/usr/lib/clc/gfx900-amdgcn--.bc`

nazar-pc committed:
Print unknown error code

nazar-pc commented:
Looks like an upstream issue with dependencies in Ubuntu: https://bugs.launchpad.net/libclc/+bug/1946381
But when I forcefully install `libclc-amdgcn` (for my GPU) I get this:

<details>

```
     Running tests/integration/main.rs (target/debug/deps/integration-c486ac06b9feced9)

running 7 tests
test opencl::test_big_random_piece ... ignored
test opencl::test_big_random_piece_pinned ... ignored
test cpu::test_known_piece ... ok
test cpu::test_random_piece ... ok

Error while building OpenCL program:
input.cl:13:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:22:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:30:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:35:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:48:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:69:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:91:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:107:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:162:5: error: use of undeclared identifier 'u15421'
<command line>:2:15: note: expanded from here
input.cl:165:36: error: use of undeclared identifier 'junk'
input.cl:316:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:434:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:445:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:520:21: warning: initializing '__global uchar *__private' (aka '__global unsigned char *__private') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]

thread 'opencl::test_random_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:31:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test opencl::test_random_piece ... FAILED

Error while building OpenCL program:
input.cl:13:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:22:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:30:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:35:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:48:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:69:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:91:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:107:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:162:5: error: use of undeclared identifier 'u15421'
<command line>:2:15: note: expanded from here
input.cl:165:36: error: use of undeclared identifier 'junk'
input.cl:316:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:434:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:445:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:520:21: warning: initializing '__global uchar *__private' (aka '__global unsigned char *__private') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]

thread 'opencl::test_known_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:55:42
test opencl::test_known_piece ... FAILED

Error while building OpenCL program:
input.cl:13:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:22:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:30:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:35:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:48:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:69:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:91:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:107:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:162:5: error: use of undeclared identifier 'u15421'
<command line>:2:15: note: expanded from here
input.cl:165:36: error: use of undeclared identifier 'junk'
input.cl:316:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:434:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:445:1: error: OpenCL C version 1.1 does not support the 'static' storage class specifier
input.cl:520:21: warning: initializing '__global uchar *__private' (aka '__global unsigned char *__private') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]

thread 'opencl::test_random_piece_pinned' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:72:42
test opencl::test_random_piece_pinned ... FAILED

failures:

failures:
    opencl::test_known_piece
    opencl::test_random_piece
    opencl::test_random_piece_pinned

test result: FAILED. 2 passed; 3 failed; 2 ignored; 0 measured; 0 filtered out; finished in 0.23s

error: test failed, to rerun pass '--test integration'
```

</details>

So, that Open Source abstraction only supports OpenCL 1.1. I'm wondering which platform with Open Source drives did you test this on @simonatsn? `clinfo` output would help and required dependencies as well. So far it doesn't seem like it is even capable of that :disappointed:

nazar-pc commented:
Forcefully overriding OpenCL version (it seems a lot of OpenCL 1.2 features are actually supported, so I gave it a try) with `CLOVER_PLATFORM_VERSION_OVERRIDE=1.2 CLOVER_DEVICE_VERSION_OVERRIDE=1.2 CLOVER_DEVICE_CLC_VERSION_OVERRIDE=1.2 cargo test --features=opencl -- --nocapture` results in this:

<details>

```
     Running tests/integration/main.rs (target/debug/deps/integration-c486ac06b9feced9)

running 7 tests
test opencl::test_big_random_piece ... ignored
test opencl::test_big_random_piece_pinned ... ignored
test cpu::test_known_piece ... ok
test cpu::test_random_piece ... ok

Error while building OpenCL program:
input.cl:162:5: error: use of undeclared identifier 'u15421'
<command line>:2:15: note: expanded from here
input.cl:165:36: error: use of undeclared identifier 'junk'
input.cl:520:21: warning: initializing '__global uchar *__private' (aka '__global unsigned char *__private') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]

thread 'opencl::test_random_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:31:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test opencl::test_random_piece ... FAILED

Error while building OpenCL program:
input.cl:162:5: error: use of undeclared identifier 'u15421'
<command line>:2:15: note: expanded from here
input.cl:165:36: error: use of undeclared identifier 'junk'
input.cl:520:21: warning: initializing '__global uchar *__private' (aka '__global unsigned char *__private') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]

thread 'opencl::test_known_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:55:42
test opencl::test_known_piece ... FAILED

Error while building OpenCL program:
input.cl:162:5: error: use of undeclared identifier 'u15421'
<command line>:2:15: note: expanded from here
input.cl:165:36: error: use of undeclared identifier 'junk'
input.cl:520:21: warning: initializing '__global uchar *__private' (aka '__global unsigned char *__private') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers [-Wincompatible-pointer-types-discards-qualifiers]

thread 'opencl::test_random_piece_pinned' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests/integration/opencl.rs:72:42
test opencl::test_random_piece_pinned ... FAILED

failures:

failures:
    opencl::test_known_piece
    opencl::test_random_piece
    opencl::test_random_piece_pinned

test result: FAILED. 2 passed; 3 failed; 2 ignored; 0 measured; 0 filtered out; finished in 0.21s
```

</details>

ozgunozerk commented:
Windows error evolved into this (tried with both `u32` and `VCC_T`):
<details>
C:\Users\ozgun\sloth256-189>cargo test --features=opencl -- --nocapture
   Compiling sloth256-189 v0.2.2 (C:\Users\ozgun\sloth256-189)
    Finished test [unoptimized + debuginfo] target(s) in 4.28s
     Running unittests (target\debug\deps\sloth256_189-a6ffb2d3046a543d.exe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests\integration\main.rs (target\debug\deps\integration-77219c89e84110ab.exe)

running 7 tests
test opencl::test_big_random_piece ... ignored
test opencl::test_big_random_piece_pinned ... ignored
test cpu::test_known_piece ... ok
test cpu::test_random_piece ... ok

Error while building OpenCL program:
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:153:5: error: implicit declaration of function '__asm__' is invalid in C99
    __asm__("v_mul_lo_u32 %0, %1, %2"
    ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:154:18: error: expected ')'
                 : "=v"(z.w[0]) : "v"(x), "v"(y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:153:12: note: to match this '('
    __asm__("v_mul_lo_u32 %0, %1, %2"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:156:18: error: expected ')'
                 : "=v"(z.w[1]) : "v"(x), "v"(y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:155:12: note: to match this '('
    __asm__("v_mul_hi_u32 %0, %1, %2"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:160:12: error: overloaded function 'mad' must have the 'overloadable' attribute
inline u64 mad(u64 z, u32 x, u32 y)
           ^
c:\constructicon\builds\gfx\seven\21.30\stream\opencl\compiler\clc2\ocl-headers\build\wNow64a\B_rel\opencl12_builtins.h:11680:49: note: previous overload of function is here
half16 const_func __attribute__((overloadable)) mad(half16 a, half16 b, half16 c);
                                                ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:165:18: error: expected ')'
                 : "+v" (z), "=s" (junk) : "v" (x), "v" (y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:164:12: note: to match this '('
    __asm__("v_mad_u64_u32 %0, %1, %2, %3, %0"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T1.cl:520:21: warning: initializing '__global uchar *' (aka '__global unsigned char *') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers
    __global uchar* iv = _iv + x * 32;
                    ^    ~~~~~~~~~~~~
1 warning and 5 errors generated.

error: Clang front-end compilation failed!
Frontend phase failed compilation.
Error: Compiling CL to IR

thread 'opencl::test_random_piece_pinned' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests\integration\opencl.rs:72:42
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test opencl::test_random_piece_pinned ... FAILED

Error while building OpenCL program:
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:153:5: error: implicit declaration of function '__asm__' is invalid in C99
    __asm__("v_mul_lo_u32 %0, %1, %2"
    ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:154:18: error: expected ')'
                 : "=v"(z.w[0]) : "v"(x), "v"(y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:153:12: note: to match this '('
    __asm__("v_mul_lo_u32 %0, %1, %2"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:156:18: error: expected ')'
                 : "=v"(z.w[1]) : "v"(x), "v"(y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:155:12: note: to match this '('
    __asm__("v_mul_hi_u32 %0, %1, %2"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:160:12: error: overloaded function 'mad' must have the 'overloadable' attribute
inline u64 mad(u64 z, u32 x, u32 y)
           ^
c:\constructicon\builds\gfx\seven\21.30\stream\opencl\compiler\clc2\ocl-headers\build\wNow64a\B_rel\opencl12_builtins.h:11680:49: note: previous overload of function is here
half16 const_func __attribute__((overloadable)) mad(half16 a, half16 b, half16 c);
                                                ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:165:18: error: expected ')'
                 : "+v" (z), "=s" (junk) : "v" (x), "v" (y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:164:12: note: to match this '('
    __asm__("v_mad_u64_u32 %0, %1, %2, %3, %0"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T3.cl:520:21: warning: initializing '__global uchar *' (aka '__global unsigned char *') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers
    __global uchar* iv = _iv + x * 32;
                    ^    ~~~~~~~~~~~~
1 warning and 5 errors generated.

error: Clang front-end compilation failed!
Frontend phase failed compilation.
Error: Compiling CL to IR

thread 'opencl::test_known_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests\integration\opencl.rs:55:42
test opencl::test_known_piece ... FAILED

Error while building OpenCL program:
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:153:5: error: implicit declaration of function '__asm__' is invalid in C99
    __asm__("v_mul_lo_u32 %0, %1, %2"
    ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:154:18: error: expected ')'
                 : "=v"(z.w[0]) : "v"(x), "v"(y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:153:12: note: to match this '('
    __asm__("v_mul_lo_u32 %0, %1, %2"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:156:18: error: expected ')'
                 : "=v"(z.w[1]) : "v"(x), "v"(y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:155:12: note: to match this '('
    __asm__("v_mul_hi_u32 %0, %1, %2"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:160:12: error: overloaded function 'mad' must have the 'overloadable' attribute
inline u64 mad(u64 z, u32 x, u32 y)
           ^
c:\constructicon\builds\gfx\seven\21.30\stream\opencl\compiler\clc2\ocl-headers\build\wNow64a\B_rel\opencl12_builtins.h:11680:49: note: previous overload of function is here
half16 const_func __attribute__((overloadable)) mad(half16 a, half16 b, half16 c);
                                                ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:165:18: error: expected ')'
                 : "+v" (z), "=s" (junk) : "v" (x), "v" (y));
                 ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:164:12: note: to match this '('
    __asm__("v_mad_u64_u32 %0, %1, %2, %3, %0"
           ^
C:\Users\ozgun\AppData\Local\Temp\\OCL5816T5.cl:520:21: warning: initializing '__global uchar *' (aka '__global unsigned char *') with an expression of type 'const __global uchar *' (aka 'const __global unsigned char *') discards qualifiers
    __global uchar* iv = _iv + x * 32;
                    ^    ~~~~~~~~~~~~
1 warning and 5 errors generated.

error: Clang front-end compilation failed!
Frontend phase failed compilation.
Error: Compiling CL to IR

thread 'opencl::test_random_piece' panicked at 'called `Result::unwrap()` on an `Err` value: OpenCLError("CL_BUILD_PROGRAM_FAILURE")', tests\integration\opencl.rs:31:42
test opencl::test_random_piece ... FAILED

failures:

failures:
    opencl::test_known_piece
    opencl::test_random_piece
    opencl::test_random_piece_pinned

test result: FAILED. 2 passed; 3 failed; 2 ignored; 0 measured; 0 filtered out; finished in 0.43s

error: test failed, to rerun pass '--test integration'
</details>

nazar-pc committed:
Fix typos

nazar-pc committed:
Fix GPU detection

nazar-pc committed:
Fix clippy complaint

nazar-pc committed:
Add function to check whether pinned memory is supported at all, use it in tests

nazar-pc committed:
Fix clippy warning

nazar-pc committed:
Fix typo

nazar-pc committed:
Fix AMD checks for Open Source OpenCL driver (not supported)

nazar-pc committed:
Improve error codes for device checks

nazar-pc commented:
While it doesn't work with Open Source AMD driver under Linux, it works just fine with ROCm-based proprietary driver.

I adjusted code to skip AMD devices that are using Open Source IDC.

nazar-pc commented on src/opencl/non_nvidia.cl:
```diff
@@ -0,0 +1,489 @@
+inline u64 mul(u32 x, u32 y)
+{
+    union { u64 W; u32 w[2]; } z;
+    __asm__("v_mul_lo_u32 %0, %1, %2"
```
@simonatsn we have faced an issue on Windows where this may not compile, see https://github.com/subspace/sloth256-189/issues/29

I'm wondering if these ASM parts are mandatory or we can fall back to the same code as Intel GPUs use above if `__asm__` macro is not available?

simonatsn replied:
Mostly likely the code for the Intel GPU won't work. In the past we had trouble with the AMD compiler producing incorrect results for that code, possibly due to incorrect optimizations. 

However I expect that code should work on Windows. Is it possible the version of opencl being used is old? Can you obtain the results of clinfo? It should be using opencl2 or newer.

simonatsn replied:
I think that makes sense (filtering out particular GPUs if they are problematic), especially for the smaller GPUs that don't contribute much to overall performance. 

Another thing to consider adding is a version check to make sure the OpenCL driver is recent. You should be able to call clGetPlatformInfo to get the CL_PLATFORM_VERSION param and make sure it's >= 2.

nazar-pc replied:
Did that in https://github.com/subspace/sloth256-189/pull/30, thanks for suggestion!
